### ISSUE TITLE
Clickhouse dialect does not support lambda expressions

### ISSUE THREAD

**Issue by simpl1g** [sqlfluff/sqlfluff#4254]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

ClickHouse DB allows to pass lambdas to work with arrays. I was trying to use them with `sqlfluff` and it does not work

Example SQL for tests
```sql
SELECT arrayFirst(x -> x = 2, [1, 1, 2, 2])
```

### Expected Behaviour

This SQL should be valid and should not raise errors

### Observed Behaviour

sqlfluff failed to parse this valid sql

```
SELECT arrayFirst(x -> x = 2, [1, 1, 2, 2])
```

```
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'SELECT'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                whitespace:                                   ' '
[L:  1, P:  8]      |                select_clause_element:
[L:  1, P:  8]      |                    function:
[L:  1, P:  8]      |                        function_name:
[L:  1, P:  8]      |                            function_name_identifier:         'arrayFirst'
[L:  1, P: 18]      |                        bracketed:
[L:  1, P: 18]      |                            start_bracket:                    '('
[L:  1, P: 19]      |                            [META] indent:
[L:  1, P: 19]      |                            expression:
[L:  1, P: 19]      |                                column_reference:
[L:  1, P: 19]      |                                    naked_identifier:         'x'
[L:  1, P: 20]      |                            unparsable:                       !! Expected: 'Nothing...'
[L:  1, P: 20]      |                                whitespace:                   ' '
[L:  1, P: 21]      |                                raw:                          '-'
[L:  1, P: 22]      |                                raw:                          '>'
[L:  1, P: 23]      |                                whitespace:                   ' '
[L:  1, P: 24]      |                                raw:                          'x'
[L:  1, P: 25]      |                                whitespace:                   ' '
[L:  1, P: 26]      |                                raw:                          '='
[L:  1, P: 27]      |                                whitespace:                   ' '
[L:  1, P: 28]      |                                numeric_literal:              '2'
[L:  1, P: 29]      |                                comma:                        ','
[L:  1, P: 30]      |                                whitespace:                   ' '
[L:  1, P: 31]      |                                start_square_bracket:         '['
[L:  1, P: 32]      |                                numeric_literal:              '1'
[L:  1, P: 33]      |                                comma:                        ','
[L:  1, P: 34]      |                                whitespace:                   ' '
[L:  1, P: 35]      |                                numeric_literal:              '1'
[L:  1, P: 36]      |                                comma:                        ','
[L:  1, P: 37]      |                                whitespace:                   ' '
[L:  1, P: 38]      |                                numeric_literal:              '2'
[L:  1, P: 39]      |                                comma:                        ','
[L:  1, P: 40]      |                                whitespace:                   ' '
[L:  1, P: 41]      |                                numeric_literal:              '2'
[L:  1, P: 42]      |                                end_square_bracket:           ']'
[L:  1, P: 43]      |                            [META] dedent:
[L:  1, P: 43]      |                            end_bracket:                      ')'
[L:  1, P: 44]      |            [META] dedent:
[L:  1, P: 44]      |    [META] end_of_file:

==== parsing violations ====
L:   1 | P:  20 |  PRS | Line 1, Position 20: Found unparsable section: ' -> x = 2, [1, 1,
                       | 2, 2]'```

### How to reproduce

```sql
-- snippet.sql
SELECT arrayFirst(x -> x = 2, [1, 1, 2, 2])
```

```shell
sqlfluff parse snippet.sql -d clickhouse -t jinja
```

### Dialect

ClickHouse

### Version

1.4.5

### Configuration

[sqlfluff]
dialect = clickhouse
exclude_rules = L014, L030, L029, L031, L032, L011, L025

[sqlfluff:rules]
tab_space_size = 2
max_line_length = 120

[sqlfluff:rules:L010]
# Keywords
capitalisation_policy = upper

[sqlfluff:rules:L040]
# Null & Boolean Literals
capitalisation_policy = upper
[sqlfluff:rules:L028]
single_table_references=unqualified
[sqlfluff:rules:L057]
# Special Characters in Identifiers
allow_space_in_identifier = False
additional_allowed_characters = ['.']

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
